### PR TITLE
Add short_text support

### PR DESCRIPTION
**Pull request by GladOSkar**
This is mostly a port from `MaxVerevkin/swaystatus@b26224d` and based on the discussions from #1205.

It works quite well for me.

Things that could be improved are:

- [x] ~~The `[block.format]` section *requires* the `full` field to be set and i couldn't figure out how to make `serde` fall back to the default if it's not given.~~
- [x] ~~The `pacman` block doesn't support the `[block.format]` section yet because it has some weird internal logic that requires direct access to the format string (which can't be accessed from the `FormatTemplate`). Possible fixes include either adding the format string to the `FormatTemplate` or reworking the way `pacman` figures out which package managers to monitor.~~
- `RotatingTextWidget`s can't be shortened from the format string, the interface doesn't really facilitate that. This only impacts the `music` block afaik. Something for a future PR imo.

Feedback + testing appreciated :)

### CODE REVIEW

File: src/blocks/bluetooth.rs [hunk 1]
```diff
@@ -224,32 +224,27 @@ pub struct Bluetooth {
...
-    fn default_format() -> String {
-        "{label} {percentage}".into()
-    }
-
-    fn default_format_unavailable() -> String {
-        "{label} x".into()
+impl Default for BluetoothConfig {
+    fn default() -> Self {
+        Self {
+            mac: String::new(),
```
Comment (MaxVerevkin) [inline src/blocks/bluetooth.rs:242]: Why? This makes `mac` optional.

File: src/formatting.rs [hunk 2]
```diff
@@ -52,7 +58,21 @@ impl FormatTemplate {
         false
     }
 
-    pub fn from_string(s: &str) -> Result<Self> {
+    // TODO: Remove?
```
Comment (MaxVerevkin) [inline src/formatting.rs:61]: Only `xrandr` uses it, right?

Comment (GladOSkar) [reply src/formatting.rs:61]: And `pacman` but that might change. I can replave both of them

File: src/blocks/bluetooth.rs (hunk 1 above)
Comment (GladOSkar) [reply src/blocks/bluetooth.rs]: Oops yes that was stupid

File: src/blocks/bluetooth.rs [hunk 3]
```diff
@@ -224,27 +224,32 @@ pub struct Bluetooth {
...
 pub struct BluetoothConfig {
     pub mac: String,
     //DEPRECATED
     //TODO remove
     pub label: Option<String>,
+    #[serde(default = "BluetoothConfig::default_hide_disconnected")]
     pub hide_disconnected: bool,
+    #[serde(default = "BluetoothConfig::default_format")]
     pub format: Option<FormatTemplate>,
+    #[serde(default = "BluetoothConfig::default_format_unavailable")]
```
Comment (MaxVerevkin) [inline src/blocks/bluetooth.rs:238]: By the way, `Option<T>` implements Default that returns `None`, so no need to to create an explicit function for that. `#[serde(default)]` should work.

Comment (MaxVerevkin) [reply src/blocks/bluetooth.rs:238]: Same for `bool`

Comment (GladOSkar) [reply src/blocks/bluetooth.rs:238]: Oh neat, thanks!

**Comment by MaxVerevkin** [comment]
> The pacman block doesn't support the [block.format] section yet because it has some weird internal logic that requires direct access to the format string

To be honest, `pacman` block seems over-complicated for some reason. I think it's better to simplify it somehow not to rely on text of format string.

**Comment by MaxVerevkin** [comment]
> The [block.format] section requires the full field to be set and i couldn't figure out how to make serde fall back to the default if it's not given.

Perhaps turn`FormatTemplate`'s `full` field into an `Option` and modify the `default_format!` macro to take this into account? And let `FormatTemplate::render` return `""` if `full` is None?

File: src/blocks/pacman.rs [hunk 4]
```diff
@@ -142,14 +142,13 @@ impl ConfigBlock for Pacman {
...
-            format_singular: FormatTemplate::from_string(&block_config.format_singular)
-                .block_error(
-                    "pacman",
-                    "Invalid format specified for pacman::format_singular",
-                )?,
-            format_up_to_date: FormatTemplate::from_string(&block_config.format_up_to_date)
+            format_singular: FormatTemplate::new(&block_config.format_singular, None).block_error(
+                "pacman",
+                "Invalid format specified for pacman::format_singular",
+            )?,
```
Comment (MaxVerevkin) [inline src/blocks/pacman.rs:150]: Off topic but you can remove those `.block_error()`s to be consistent with other blocks.

**Comment by GladOSkar** [comment]
> Perhaps turnFormatTemplate's full field into an Option and modify the default_format! macro to take this into account? And let FormatTemplate::render return "" if full is None?

Good idea, i did basically that but then decided to change a bunch of other things because doing that within the macro was quite cumbersome and also the `FormatTemplate` in the BlockConfigs don't need to be Options anymore then. Plus it made more sense to me this way. But it works :)

I also updated the `pacman` block now in a way that i think should work, but since i don't have arch I'd appreciate some more thorough testing, as with the rest of this PR.

I'd consider this PR feature-complete now, feel free to let me know if you'd recommend any other changes

**Comment by MaxVerevkin** [comment]
> Edit: (Oh and GH actions seems to be complaining about clippy lints but i can only find some unrelated clippy warnings so i wouldn't include those changes here)

`formatting` module's test are not compiling.

File: src/formatting.rs [hunk 5]
```diff
@@ -39,20 +30,51 @@ fn unexpected_token<T>(token: char) -> Result<T> {
...
+    short: Option<Vec<Token>>,
+}
+
 impl FormatTemplate {
     /// Whether the format string contains given placeholder
     pub fn contains(&self, var: &str) -> bool {
-        for token in &self.tokens {
-            if let Token::Var(ref placeholder) = token {
-                if placeholder.name == var {
-                    return true;
```
Comment (MaxVerevkin) [inline src/formatting.rs:48]: I think `FormatTemplate::contains()` should take `short` into account too.

**Comment by MaxVerevkin** [comment]
> Edit: (Oh and GH actions seems to be complaining about clippy lints but i can only find some unrelated clippy warnings so i wouldn't include those changes here)

Build fails because of formatting. Try `cargo fmt`.

**Comment by ammgws** [comment]
@matclab Any comments? (since you did something similar in https://github.com/greshake/i3status-rust/pull/703, though I believe it stalled because we were waiting on someone to implement an enhancement on the i3 side)

**Comment by MaxVerevkin** [comment]
So what about `time`?

File: src/formatting.rs (hunk 5 above)
Comment (GladOSkar) [reply src/formatting.rs:48]: As in check both? I figured everything in `short` would also be in `full` but i guess that's reasonable

**Comment by GladOSkar** [comment]
> So what about time?

Not sure what exactly you mean but `time` uses a very different `format` system (directly passing it to the unix `date` binary afaik), hence this system is not suited to that. But of course it would be neat for it to be able to be shortened as well

**Comment by GladOSkar** [comment]
> Build fails because of formatting. Try cargo fmt.

Didn't help apparently. I already ran `rustfmt` and `clippy` as well. I reakky can't figure out what it's complaining about

**Comment by ammgws** [comment]
Doesn't look like any of the strftime modifiers conflict with our format string stuff (curly brackets), so don't see why time's `format` option couldn't be handled as a FormatTemplate internally?

https://man7.org/linux/man-pages/man3/strftime.3.html

**Comment by GladOSkar** [comment]
Hm i guess you're right it'd just be a static format without any placeholders then

**Comment by MaxVerevkin** [comment]
> Didn't help apparently. I already ran rustfmt and clippy as well. I really can't figure out what it's complaining about

Right now test don't compile.

**Comment by matclab** [comment]
> @matclab Any comments? (since you did something similar in #703, though I believe it stalled because we were waiting on someone to implement an enhancement on the i3 side)

i3 side support would be the butter. 
On #703, I added some user UI which allows to switch between three states (short enforced, long enforced, choosen by i3) by midlle clicking. It was somewhat usefull to me, but I'm not fond of the UX (poor discoverability and was difficult to know in which state the widget was).
Also my solution was to add specific code to each widget. Current solution seems much better.

So I think I'll give a try to this PR this evening to give you some feedbac, but it looks nice to me.

**Comment by GladOSkar** [comment]
> Right now test don't compile.

Oops

**Comment by GladOSkar** [comment]
@matclab i agree. The `i3` feature would make the functionality of this PR nicer as well but isn't required.

Being able to switch the format by clicking might be neat but also a bit overkill for now. We can maybe do that along with/after the unification of the click handlers discussed in #1038.

File: src/formatting.rs [hunk 6]
```diff
@@ -119,40 +157,86 @@ impl FormatTemplate {
...
-                                    "util",
-                                    &format!("Unknown placeholder in format string: {}", var.name),
-                                )?
-                                .format(&var)?,
-                        ),
-                )),
+impl Default for FormatTemplate {
+    fn default() -> Self {
+        Self {
+            full: None,
```
Comment (MaxVerevkin) [inline src/formatting.rs:165]: A simple `derive` is sufficient.

**Comment by MaxVerevkin** [comment]
> Doesn't look like any of the strftime modifiers conflict with our format string stuff (curly brackets), so don't see why time's format option couldn't be handled as a FormatTemplate internally?

Perhaps just add an explicit `format_short`? Or add `fn FormatTemplate::raw()` that would return the string it was constructed with?

File: src/formatting.rs (hunk 6 above)
Comment (GladOSkar) [reply src/formatting.rs]: Ah yes those are the parts of Rust that i don't have an intuition for yet

**Comment by ammgws** [comment]
What's the status on this?

**Comment by matclab** [comment]
I'm using the version built 2 days ago and it works fine to my taste.

**Comment by GladOSkar** [comment]
> What's the status on this?

Well it's complete imo and I implemented all of @MaxVerevkin's feedback. But it could probably need a bunch more testing since a lot was changed. Especially `pacman` with the various placeholders since I don't run Arch.

**Comment by MaxVerevkin** [comment]
>What's the status on this?

At the moment I'm running `swaystatus` which has exact same `short_text` management and it works fine. 

However, I can note that on the `swaybar` side, the support is kinda raw. In addition to lack of progressive collapsing, if *one* of my outputs switches to `short_text`, all the outer outputs switch too, which is sad (I have a small laptop screen and an external monitor).

> Especially `pacman` with the various placeholders since I don't run Arch.

I'll test it today (even though I don't like the idea of always seeing the number of updates (and if did, I would just use `checkupdates | wc -l`))

**Comment by ammgws** [comment]
Pacman block is displaying the respective arch and AUR update counts correctly here.

**Comment by MaxVerevkin** [comment]
@ivankovnatsky do you want to test this?

File: doc/blocks.md [hunk 7]
```diff
@@ -1724,6 +1724,18 @@ command = "echo '<b>1 &amp;</b>'"
...
+
+The field can be set as a string (`format = "{my_format}"`) or as a section:
+```toml
+[[block]]
+block = "my_block"
+[block.format]
+full = "{my_full_format}"
+short = "{my_short_format}"
+```
+Your `i3` or `sway` will switch all blocks over to the `short` variant whenever there isn't nough space on your screen for the `full` status bar.
```
Comment (ivankovnatsky) [inline doc/blocks.md:1737]:
```suggestion
 Your `i3` or `sway` will switch all blocks over to the `short` variant whenever there isn't enough space on your screen for the `full` status bar.
```

**Comment by GladOSkar** [comment]
@ammgws if `pacman` works and nobody has any issues anymore i think this is good to go

**Comment by ivankovnatsky** [comment]
guys, thanks a lot for taking time implementing this!

laptop/monitor bars running at the same time, i like it. though running in a plain theme i guess:

laptop
monitor

running my config as is gives me:

```
May 10 15:24:24 thinkpad sway[34355]: Configuration error: failed to parse TOML from file contents.
May 10 15:24:24 thinkpad sway[34355]: Cause: Icon set 'awesome5' not found for key `icons` at line 107 column 1
```

using plain theme gives the same error:

```
May 10 15:34:10 thinkpad sway[46970]: Configuration error: failed to parse TOML from file contents.
May 10 15:34:10 thinkpad sway[46970]: Cause: Theme 'plain' not found. for key `theme` at line 2 column 9
```

i was able to run i3status-rust only without `theme` and `icons` definitions.

i guess i compiled it in a weird way, i used nix overlay to built it:

```nix
{ lib, rustPlatform, fetchFromGitHub, pkg-config, makeWrapper, dbus
, libpulseaudio, notmuch, openssl, ethtool }:

rustPlatform.buildRustPackage rec {
  pname = "i3status-rust";
  version = "4bf8495dc5a8292acacc39f27114ce838a6c4183";

  src = fetchFromGitHub {
    owner = "GladOSkar";
    repo = pname;
    rev = "${version}";
    sha256 = "sha256-sgbT9vD/d/2obW+f3DcaSdkEXTlPyUFTk5wYsPASL3o=";
  };

  cargoSha256 = "sha256-+frd9sLelFLfv99nQl5bIQlVFXhkC7AP2ZD8hmPj+u0=";

  nativeBuildInputs = [ pkg-config makeWrapper ];

  buildInputs = [ dbus libpulseaudio notmuch openssl ];

  cargoBuildFlags = [ "--features=notmuch" "--features=maildir" ];

  postFixup = ''
    wrapProgram $out/bin/i3status-rs --prefix PATH : "${ethtool}/bin"
  '';

  # Currently no tests are implemented, so we avoid building the package twice
  doCheck = false;

  meta = with lib; {
    description =
      "Very resource-friendly and feature-rich replacement for i3status";
    homepage = "https://github.com/greshake/i3status-rust";
    license = licenses.gpl3;
    maintainers = with maintainers; [ backuitist globin ma27 ];
    platforms = platforms.linux;
  };
}
```

Anyway, the hiding is works as expected.

**Comment by ivankovnatsky** [comment]
@GladOSkar yeah, i guess, i rebuilt your current change and restarted sway all the way, and the icons appeared as expected.

**Comment by ivankovnatsky** [comment]
bigger screen:
image

smaller screen:
image

ignore mem widgets, didn't formatted yet.

```toml
theme = "space-villain"
icons = "awesome5"

[[block]]
block = "custom"
command = "echo  $(uname -r)"
interval = "once"

[[block]]
block = "custom"
command = "[ $(nixos-version --revision) != $(curl -s https://api.github.com/repos/NixOS/nixpkgs/git/refs/heads/nixos-unstable | jq '.object.sha' -r ) ] && echo '{\"state\":\"Info\", \"text\": \"Update\"}' || echo '{\"state\":\"Idle\", \"text\": \"No Update\"}'"
interval = 1800
json = true

[[block]]
block = "cpu"
interval = 1

[block.format]
full = "{utilization} {frequency}"
short = ""

[[block]]
block = "load"
interval = 1

[block.format]
full = "{1m} {5m} {15m}"
short = ""

[[block]]
block = "temperature"
chip = "*-isa-*"
collapsed = false
inputs = ["temp1"]
interval = 1

[block.format]
full = "{max}"
short = ""

[[block]]
block = "custom"
command = "echo  $(cat /sys/class/hwmon/hwmon4/fan1_input) RPM"
interval = 1

[[block]]
block = "memory"
display_type = "memory"

[[block]]
block = "memory"
display_type = "swap"

[[block]]
alert = 200
alias = "/"
block = "disk_space"
info_type = "used"
path = "/"
warning = 150

[block.format]
full = " {used}/{total}"
short = ""

[[block]]
block = "net"
device = "wlp2s0"
hide_inactive = true
interval = 1

[block.format]
full = "{speed_down} {speed_up}"
short = ""

[[block]]
block = "battery"
driver = "upower"

[block.format]
full = "{percentage}"
short = ""

[[block]]
block = "keyboard_layout"
driver = "sway"
format = " {layout}"

[[block]]
block = "sound"
on_click = "pavucontrol"

[[block]]
block = "weather"

[block.format]
full = "{temp} {apparent}  {humidity} 煮 {wind_kmh} km/h"
short = ""

[block.service]
api_key = ""
city_id = ""
name = "openweathermap"
units = "metric"

[[block]]
block = "time"
format = "%a %b %d %H:%M"
```

**Comment by GladOSkar** [comment]
Hm. I'll have another look at that

**Comment by ivankovnatsky** [comment]
i think keeping the widget icon is good idea (in short version), since it could be highlighted with color, we might not see the details, but the color will tell in general.

going back to what i imagined this implementation is much-much better!

**Comment by MaxVerevkin** [comment]
> bigger screen:

>smaller screen:

Do you use `i3` or `sway`? for me on `sway`, all outputs have the same, either short or full text.

**Comment by GladOSkar** [comment]
On my `i3` it is able to keep them separate, `full` on a bar with enough space and `short` on one without.

**Comment by ivankovnatsky** [comment]
> Do you use i3 or sway? for me on sway, all outputs have the same, either short or full text.

sway, i meant that on a bigger screen all my widgets get it's space, on a laptop it will fallback to short versions, where applicable and configured.

if i deliberately clone some widgets to fill the width it will fallback to short versions on both screens:

screen with longer width: 
image

shorter width: 
image

**Comment by MaxVerevkin** [comment]
OK, so it's `swaybar`'s bug.

**Comment by GladOSkar** [comment]
@ammgws any blockers left?

**Pull request merged**